libbpf-tools/biosnoop.bpf.c

```diff
@@ -0,0 +1,153 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Wenbo Zhang
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include "biosnoop.h"
+
+#define MAX_ENTRIES 10240
+
+const volatile char targ_disk[DISK_NAME_LEN] = {};
+const volatile bool targ_queued = false;
+
+struct piddata {
+	char comm[TASK_COMM_LEN];
+	u32 pid;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct request *);
+	__type(value, struct piddata);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} infobyreq SEC(".maps");
+
+struct stage {
+	u64 insert;
+	u64 issue;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct request *);
+	__type(value, struct stage);
+} start SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(u32));
+} events SEC(".maps");
+
+SEC("fentry/blk_account_io_start")
+int BPF_PROG(tp_btf__blk_account_io_start, struct request *rq, bool new_io)
```

Comment:
not really tp_btf

also, please double-check that https://github.com/torvalds/linux/commit/b5af37ab3a2b143e278340d2c6fa5790d53817e7 doesn't break the logic, seems like there were some refactorings recently.

Reply:
> not really tp_btf
Oh sorry, I misspelled it.

> also, please double-check that torvalds/linux@b5af37a doesn't break the logic, seems like there were some refactorings recently.

Thank you @anakryiko , I found that for the same rq,  `blk_account_io_start` and `blk_account_io_merge_bio` may occur on different threads. So I think I need to trace `blk_account_io_merge_bio` to update owner.